In [1]:
import pandas as pd
pd.options.display.max_columns = 99

In [2]:
loans = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', nrows=5)
loans

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501  1296599.0     5000.0       5000.0           4975.0   36 months   
1  1077430  1314167.0     2500.0       2500.0           2500.0   60 months   
2  1077175  1313524.0     2400.0       2400.0           2400.0   36 months   
3  1076863  1277178.0    10000.0      10000.0          10000.0   36 months   
4  1075358  1311748.0     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  \
0           RENT     24000.0            Verified  Dec-2011   Fully Paid   
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off   
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid   
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid   
4           RENT     80000.0     Source Verified  Dec-2011      Current   

  pymnt_plan         purpose                 title zip_code addr_state    dti  \
0          n     credit_card              Computer    860xx         AZ  27.65   
1          n             car                  bike    309xx         GA   1.00   
2          n  small_business  real estate business    606xx         IL   8.72   
3          n           other              personel    917xx         CA  20.00   
4          n           other              Personal    972xx         OR  17.94   

   delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  \
0          0.0         Jan-1985             1.0       3.0      0.0    13648.0   
1          0.0         Apr-1999             5.0       3.0      0.0     1687.0   
2          0.0         Nov-2001             2.0       2.0      0.0     2956.0   
3          0.0         Feb-1996             1.0      10.0      0.0     5598.0   
4          0.0         Jan-1996             0.0      15.0      0.0    27783.0   

  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
0      83.7%        9.0                   f       0.00           0.00   
1       9.4%        4.0                   f       0.00           0.00   
2      98.5%       10.0                   f       0.00           0.00   
3        21%       37.0                   f       0.00           0.00   
4      53.9%       38.0                   f     461.73         461.73   

    total_pymnt  total_pymnt_inv  total_rec_prncp  total_rec_int  \
0   5863.155187          5833.84          5000.00         863.16   
1   1008.710000          1008.71           456.46         435.17   
2   3005.666844          3005.67          2400.00         605.67   
3  12231.890000         12231.89         10000.00        2214.92   
4   3581.120000          3581.12          2538.27        1042.85   

   total_rec_late_fee  recoveries  collection_recovery_fee last_pymnt_d  \
0                0.00        0.00                     0.00     Jan-2015   
1                0.00      117.08                     1.11     Apr-2013   
2                0.00        0.00                     0.00     Jun-2014   
3               16.97        0.00                     0.00     Jan-2015   
4                0.00        0.00                     0.00     Jun-2016   

   last_pymnt_amnt last_credit_pull_d  collections_12_mths_ex_med  \
0           171.62           Jun-2016                         0.0   
1           119.66           Sep-2013                         0.0   
2           649.91           Jun-2016                         0.0   
3           357.48           Apr-2016                     

## Selecting the chunksize number

In [3]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)
rows=0
for chunk in chunk_iter:
    rows += len(chunk)
print(rows)

42538


In [4]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)
i = 1
for chunk in chunk_iter:
    print('chunk_{} :'.format(i), chunk.memory_usage(deep=True).sum()/(1024*1024))
    i+=1

chunk_1 : 4.580394744873047
chunk_2 : 4.576141357421875
chunk_3 : 4.577898979187012
chunk_4 : 4.579251289367676
chunk_5 : 4.575444221496582
chunk_6 : 4.577326774597168
chunk_7 : 4.575918197631836
chunk_8 : 4.578287124633789
chunk_9 : 4.576413154602051
chunk_10 : 4.57646369934082
chunk_11 : 4.589176177978516
chunk_12 : 4.588043212890625
chunk_13 : 4.594850540161133
chunk_14 : 4.828314781188965
chunk_15 : 0.868586540222168


We will choose 3000 as the size of the chunk so that it does not overrun five megabytes of memory usage.

# Exploring the Data in Chunks

### How many columns have a numeric type? How many have a string type?

In [5]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)
num = []
obj = []
for chunk in chunk_iter:
    numeric_cols = [col for col in chunk.columns if chunk[col].dtype != 'O']
    num.append(len(numeric_cols))
    object_cols = [col for col in chunk.columns if chunk[col].dtype == 'O']
    obj.append(len(object_cols))
print('Number of num columns :', num)
print('Number of obj columns :', obj)

Number of num columns : [31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 30, 30]
Number of obj columns : [21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 21, 22, 22]


In [6]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)
obj=[]
i=0
for chunk in chunk_iter:
    i+=1
    object_cols = [col for col in chunk.columns if chunk[col].dtype == 'O']
    if len(obj)>0:
        if obj != object_cols:
            print('-'*10, ' Chunk_{} '.format(i),'-'*10)
            print(object_cols)
    else:
        obj = object_cols
        print(object_cols)  
    

['term', 'int_rate', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'purpose', 'title', 'zip_code', 'addr_state', 'earliest_cr_line', 'revol_util', 'initial_list_status', 'last_pymnt_d', 'last_credit_pull_d', 'application_type']
----------  Chunk_14  ----------
['id', 'term', 'int_rate', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'purpose', 'title', 'zip_code', 'addr_state', 'earliest_cr_line', 'revol_util', 'initial_list_status', 'last_pymnt_d', 'last_credit_pull_d', 'application_type']
----------  Chunk_15  ----------
['id', 'term', 'int_rate', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'purpose', 'title', 'zip_code', 'addr_state', 'earliest_cr_line', 'revol_util', 'initial_list_status', 'last_pymnt_d', 'last_credit_pull_d', 'appli

It seems that the column ` id ` became an object columns but since it have no use for analysis, visualisation and prediction we can ignore it.

### How many unique values are there in each string column?

In [7]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)
uniques = dict()
for chunk in chunk_iter:
    string_df = chunk.select_dtypes(include=['object'])
    string_col = string_df.columns
    for col in string_col:
        val_counts = string_df[col].value_counts()
        if col in uniques:
            uniques[col].append(val_counts)
        else:
            uniques[col] = [val_counts]

            
uniques_combined = {}

for col in uniques:
    u_concat = pd.concat(uniques[col])
    u_group = u_concat.groupby(u_concat.index).sum()
    uniques_combined[col] = u_group
    print(col, u_group.shape[0])


term 2
int_rate 394
grade 7
sub_grade 35
emp_title 30658
emp_length 11
home_ownership 5
verification_status 3
issue_d 55
loan_status 9
pymnt_plan 2
purpose 14
title 21264
zip_code 837
addr_state 50
earliest_cr_line 530
revol_util 1119
initial_list_status 1
last_pymnt_d 103
last_credit_pull_d 108
application_type 1
id 3538


### Which float columns have missing values?

In [8]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)

missing = []
for chunk in chunk_iter:
    floats = chunk.select_dtypes(include=['float'])
    missing.append(floats.apply(pd.isnull).sum())

combined_missing = pd.concat(missing)
combined_missing.groupby(combined_missing.index).sum().sort_values()

member_id                        3
total_rec_int                    3
total_pymnt_inv                  3
total_pymnt                      3
revol_bal                        3
recoveries                       3
policy_code                      3
out_prncp_inv                    3
out_prncp                        3
total_rec_late_fee               3
loan_amnt                        3
last_pymnt_amnt                  3
total_rec_prncp                  3
funded_amnt_inv                  3
funded_amnt                      3
dti                              3
collection_recovery_fee          3
installment                      3
annual_inc                       7
inq_last_6mths                  32
total_acc                       32
delinq_2yrs                     32
pub_rec                         32
delinq_amnt                     32
open_acc                        32
acc_now_delinq                  32
tax_liens                      108
collections_12_mths_ex_med     148
chargeoff_within_12_

### Calculate the total memory usage across all of the chunks.

In [9]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)
memory = 0
for chunk in chunk_iter:
    memory += chunk.memory_usage(deep=True).sum()/(1024*1024)
print('total memory is {}'.format(memory))

total memory is 65.24251079559326


# Optimizing String Columns

In [10]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)
uniques = dict()
for chunk in chunk_iter:
    string_df = chunk.select_dtypes(include=['object'])
    string_col = string_df.columns
    for col in string_col:
        val_counts = string_df[col].value_counts(normalize=True)
#         print(val_counts)
        if val_counts[0]<0.005:
            if col in uniques:
                uniques[col].append(val_counts)
            else:
                uniques[col] = [val_counts]
for item in uniques:
    print(item)

emp_title
id


In [11]:
useful_obj_cols = ['term', 'sub_grade', 'home_ownership', 'verification_status', 'issue_d', 'purpose', 
                   'earliest_cr_line', 'revol_util', 'last_pymnt_d', 'last_credit_pull_d']

In [12]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', chunksize=3000)
str_cols_vc = {}
for chunk in chunk_iter:
    str_cols = chunk.select_dtypes(include=['object'])
    for col in str_cols.columns:
        current_col_vc = str_cols[col].value_counts()
        if col in str_cols_vc:
            str_cols_vc[col].append(current_col_vc)
        else:
            str_cols_vc[col] = [current_col_vc]
str_cols_vc['term']

[ 36 months    2060
  60 months     940
 Name: term, dtype: int64,
  36 months    1950
  60 months    1050
 Name: term, dtype: int64,
  36 months    1970
  60 months    1030
 Name: term, dtype: int64,
  36 months    2015
  60 months     985
 Name: term, dtype: int64,
  36 months    2002
  60 months     998
 Name: term, dtype: int64,
  36 months    1734
  60 months    1266
 Name: term, dtype: int64,
  36 months    1863
  60 months    1137
 Name: term, dtype: int64,
  36 months    2027
  60 months     973
 Name: term, dtype: int64,
  36 months    1831
  60 months    1169
 Name: term, dtype: int64,
  36 months    1866
  60 months    1134
 Name: term, dtype: int64,
  36 months    2992
  60 months       8
 Name: term, dtype: int64,
  36 months    3000
 Name: term, dtype: int64,
  36 months    3000
 Name: term, dtype: int64,
  36 months    2688
  60 months     311
 Name: term, dtype: int64,
  36 months    536
 Name: term, dtype: int64]

In [13]:
combined_vcs = {}

for col in str_cols_vc:
    combined_vc = pd.concat(str_cols_vc[col])
    final_vc = combined_vc.groupby(combined_vc.index).sum()
    combined_vcs[col] = final_vc

combined_vcs['term']

 36 months    31534
 60 months    11001
Name: term, dtype: int64

In [14]:
for col in useful_obj_cols:
    print(col)
    print(combined_vcs[col])
    print("-----------")

term
 36 months    31534
 60 months    11001
Name: term, dtype: int64
-----------
sub_grade
A1    1142
A2    1520
A3    1823
A4    2905
A5    2793
B1    1882
B2    2113
B3    2997
B4    2590
B5    2807
C1    2264
C2    2157
C3    1658
C4    1370
C5    1291
D1    1053
D2    1485
D3    1322
D4    1140
D5    1016
E1     884
E2     791
E3     668
E4     552
E5     499
F1     392
F2     308
F3     236
F4     211
F5     154
G1     141
G2     107
G3      79
G4      99
G5      86
Name: sub_grade, dtype: int64
-----------
home_ownership
MORTGAGE    18959
NONE            8
OTHER         136
OWN          3251
RENT        20181
Name: home_ownership, dtype: int64
-----------
verification_status
Not Verified       18758
Source Verified    10306
Verified           13471
Name: verification_status, dtype: int64
-----------
issue_d
Apr-2008     259
Apr-2009     333
Apr-2010     912
Apr-2011    1563
Aug-2007      74
Aug-2008     100
Aug-2009     446
Aug-2010    1175
Aug-2011    1934
Dec-2007     172
Dec-

In [15]:
chunk[useful_obj_cols]

term sub_grade home_ownership verification_status   issue_d  \
42000   36 months        C2           RENT        Not Verified  Feb-2008   
42001   36 months        G2            OWN        Not Verified  Feb-2008   
42002   36 months        E4           RENT        Not Verified  Feb-2008   
42003   36 months        G4           RENT        Not Verified  Feb-2008   
42004   36 months        B3           RENT        Not Verified  Feb-2008   
...           ...       ...            ...                 ...       ...   
42533   36 months        B3           RENT        Not Verified  Jun-2007   
42534   36 months        A5           NONE        Not Verified  Jun-2007   
42535   36 months        A3       MORTGAGE        Not Verified  Jun-2007   
42536         NaN       NaN            NaN                 NaN       NaN   
42537         NaN       NaN            NaN                 NaN       NaN   

                  purpose earliest_cr_line revol_util last_pymnt_d  \
42000  debt_consolidation         Jul-2000     100.7%     Feb-2011   
42001  debt_consolidation         Mar-1989      51.9%     Nov-2008   
42002  debt_consolidation         Dec-1998      80.7%     Feb-2011   
42003  debt_consolidation         Jul-1995      57.2%     Feb-2011   
42004  debt_consolidation         Sep-2005        74%     Mar-2010   
...                   ...              ...        ...          ...   
42533               other              NaN        NaN     Jun-2010   
42534               other              NaN        NaN     Jun-2010   
42535               other              NaN        NaN     Jun-2010   
42536                 NaN              NaN        NaN          NaN   
42537                 NaN              NaN        NaN          NaN   

      last_credit_pull_d  
42000           Jun-2016  
42001           Jun-2016  
42002           Jun-2016  
42003           Jun-2011  
42004           Aug-2010  
...                  ...  
42533           May-2007  
42534           Aug-2007  
42535           Feb-2015  
42536                NaN  
42537                NaN  

[538 rows x 10 columns]

In [16]:
convert_col_dtypes = {
    "sub_grade": "category", "home_ownership": "category", 
    "verification_status": "category", "purpose": "category"
}
date_cols = ["issue_d", "earliest_cr_line", "last_pymnt_d", "last_credit_pull_d"]

In [17]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', 
                         chunksize=3000, 
                         dtype=convert_col_dtypes, 
                         parse_dates=date_cols)
for chunk in chunk_iter:
    term_cleaned = chunk['term'].str.lstrip(" ").str.rstrip(" months")
    chunk['term'] = pd.to_numeric(term_cleaned)
    revol_cleaned = chunk['revol_util'].str.rstrip("%")
    chunk['revol_util'] = pd.to_numeric(revol_cleaned)
    
chunk.dtypes

id                                    object
member_id                            float64
loan_amnt                            float64
funded_amnt                          float64
funded_amnt_inv                      float64
term                                 float64
int_rate                              object
installment                          float64
grade                                 object
sub_grade                           category
emp_title                             object
emp_length                            object
home_ownership                      category
annual_inc                           float64
verification_status                 category
issue_d                       datetime64[ns]
loan_status                           object
pymnt_plan                            object
purpose                             category
title                                 object
zip_code                              object
addr_state                            object
dti       

In [18]:
chunk_iter = pd.read_csv('OneDrive\Documents\my_datasets\Loans_2007.csv', 
                         chunksize=3000, 
                         dtype=convert_col_dtypes, 
                         parse_dates=date_cols)
memory = 0
for chunk in chunk_iter:
    term_cleaned = chunk['term'].str.lstrip(" ").str.rstrip(" months")
    chunk['term'] = pd.to_numeric(term_cleaned)
    revol_cleaned = chunk['revol_util'].str.rstrip("%")
    chunk['revol_util'] = pd.to_numeric(revol_cleaned)
    memory += chunk.memory_usage(deep=True).sum()/(1024*1024)
print('total memory is {}'.format(memory))

total memory is 41.08849239349365


It looks like we were able to realize some powerful memory savings by converting to the category type and converting string columns to numeric ones.

65.2425 >>> 41.0885 megabytes